src/self_driving/planning/pilot.cpp

```diff
@@ -71,27 +72,40 @@ void Pilot::PerformPlanning() {
   }
 
   // Perform planning
-  std::vector<std::pair<const std::string, catalog::db_oid_t>> best_action_seq;
+  std::vector<pilot::ActionTreeNode> best_action_seq;
   Pilot::ActionSearch(&best_action_seq);
 
   metrics_thread_->ResumeMetrics();
 }
 
-void Pilot::ActionSearch(std::vector<std::pair<const std::string, catalog::db_oid_t>> *best_action_seq) {
+void Pilot::ActionSearch(std::vector<pilot::ActionTreeNode> *best_action_seq) {
   auto num_segs = forecast_->GetNumberOfSegments();
   auto end_segment_index = std::min(action_planning_horizon_ - 1, num_segs - 1);
 
   auto mcst =
       pilot::MonteCarloTreeSearch(common::ManagedPointer(this), common::ManagedPointer(forecast_), end_segment_index);
-  mcst.BestAction(simulation_number_, best_action_seq,
-                  settings_manager_->GetInt64(settings::Param::pilot_memory_constraint));
-  for (uint64_t i = 0; i < best_action_seq->size(); i++) {
+  mcst.RunSimulation(simulation_number_, settings_manager_->GetInt64(settings::Param::pilot_memory_constraint));
+
+  // Record the top 3 at each level along the "best action path".
+  // TODO(wz2): May want to improve this at a later time.
+  std::vector<std::vector<pilot::ActionTreeNode>> layered_action;
+  mcst.BestAction(&layered_action, 3);
+  for (size_t i = 0; i < layered_action.size(); i++) {
+    pilot::ActionTreeNode &atn = layered_action[i].front();
```

Comment:
nit: I prefer not to use abbreviations unless it's too long. Just use `action` here.